### PR TITLE
[HttpFoundation] Add `PRIVATE_SUBNETS` as a shortcut for private IP address ranges to `Request::setTrustedProxies()`

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q             | A
| ------------- | ---
| Branch?       | 7.2
| Bug fix?      | no
| New feature?  | no
| Deprecations? | no
| Issues        | -
| License       | MIT

Let's save some memory allocations and callbacks when we can.

Tweaks #33574 and #52924

### CODE REVIEW

**Review by derrabus — APPROVED** [review]

File: src/Symfony/Component/HttpFoundation/Request.php [hunk 1]
```diff
@@ -520,20 +520,26 @@ public function overrideGlobals(): void
      *
      * You should only list the reverse proxies that you manage directly.
      *
-     * @param array $proxies          A list of trusted proxies, the string 'REMOTE_ADDR' will be replaced with $_SERVER['REMOTE_ADDR']
-     * @param int   $trustedHeaderSet A bit field of Request::HEADER_*, to set which headers to trust from your proxies
+     * @param array                          $proxies          A list of trusted proxies, the string 'REMOTE_ADDR' will be replaced with $_SERVER['REMOTE_ADDR']
```
Comment (xabbuh) [inline src/Symfony/Component/HttpFoundation/Request.php]: should probably also mention `PRIVATE_SUBNETS`

**Review by xabbuh — APPROVED** [review]

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php [hunk 2]
```diff
@@ -113,8 +113,8 @@ public function getConfigTreeBuilder(): TreeBuilder
                 ->end()
                 ->scalarNode('trusted_proxies')
                     ->beforeNormalization()
-                        ->ifTrue(fn ($v) => 'private_ranges' === $v)
-                        ->then(fn ($v) => implode(',', IpUtils::PRIVATE_SUBNETS))
+                        ->ifTrue(fn ($v) => 'private_ranges' === $v || 'PRIVATE_SUBNETS' === $v)
+                        ->then(fn () => IpUtils::PRIVATE_SUBNETS)
```
Comment (xabbuh) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php:117]:
```suggestion
                        ->then(fn () => implode(',', IpUtils::PRIVATE_SUBNETS))
```

**Review by xabbuh — CHANGES_REQUESTED** [review]

**Review by nicolas-grekas — COMMENTED** [review]
While reading the code around, I realized that `private_ranges` for trusted proxies works only for explicit static configuration, aka env vars are not supported.
I fixed this by adding support for the `PRIVATE_SUBNETS` special value in setTrustedProxies, next to the already supported `REMOTE_ADDR`. Note the name and casing change, which look desired to me for consistency with `REMOTE_ADDR`. `private_ranges` is also supported for legacy reasons.

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php (hunk 2 above)
Comment (nicolas-grekas) [reply src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php:117]: I went with a variableNode instead, the best type for the parameter is an array, although string can be used also but require some processing

File: src/Symfony/Component/HttpFoundation/Request.php (hunk 1 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/HttpFoundation/Request.php]: updated

File: src/Symfony/Component/HttpFoundation/Request.php [hunk 3]
```diff
@@ -520,20 +520,26 @@ public function overrideGlobals(): void
...
+        if (false !== $i = array_search('REMOTE_ADDR', $proxies, true)) {
+            if (isset($_SERVER['REMOTE_ADDR'])) {
+                $proxies[$i] = $_SERVER['REMOTE_ADDR'];
+            } else {
+                unset($proxies[$i]);
+                $proxies = array_values($proxies);
             }
+        }
+
+        if (false !== ($i = array_search('PRIVATE_SUBNETS', $proxies, true)) || false !== ($i = array_search('private_ranges', $proxies, true))) {
```
Comment (stof) [inline src/Symfony/Component/HttpFoundation/Request.php:537]: should `private_ranges` be deprecated in favor of `PRIVATE_SUBNETS` or no ?

Comment (nicolas-grekas) [reply src/Symfony/Component/HttpFoundation/Request.php:537]: I don't think it's worth the trouble for the community.

**Review by xabbuh — APPROVED** [review]

**Comment by ValentinRgt** [comment]
Hi, this improvement be retroactive up to and including 6.4, 7.0 and 7.1?

**Comment by xabbuh** [comment]
New features are not backported to already released Symfony versions. You will have to upgrade to 7.2 to be able to use it.

**Comment by faizanakram99** [comment]
> While reading the code around, I realized that `private_ranges` for trusted proxies works only for explicit static configuration, aka env vars are not supported. I fixed this by adding support for the `PRIVATE_SUBNETS` special value in setTrustedProxies, next to the already supported `REMOTE_ADDR`. Note the name and casing change, which look desired to me for consistency with `REMOTE_ADDR`. `private_ranges` is also supported for legacy reasons.

@nicolas-grekas sorry for the ping, will PRIVATE_SUBNETS and REMOTE_ADDR work with the new env vars SYMFONY_TRUSTED_PROXIES?

**Comment by nicolas-grekas** [comment]
They should yes.

**Pull request merged**